view/src/main/java/cmm/apps/esmorga/view/registration/RegistrationViewModel.kt

```diff
@@ -0,0 +1,113 @@
+import cmm.apps.esmorga.view.registration.model.RegistrationUiState
+import cmm.apps.esmorga.view.registration.model.RegistrationViewHelper.getEmailAlreadyInUseErrorText
+import cmm.apps.esmorga.view.registration.model.RegistrationViewHelper.getEmailErrorText
+import cmm.apps.esmorga.view.registration.model.RegistrationViewHelper.getEmptyFieldErrorText
+import cmm.apps.esmorga.view.registration.model.RegistrationViewHelper.getLastNameErrorText
+import cmm.apps.esmorga.view.registration.model.RegistrationViewHelper.getNameErrorText
+import cmm.apps.esmorga.view.registration.model.RegistrationViewHelper.getPasswordErrorText
+import cmm.apps.esmorga.view.registration.model.RegistrationViewHelper.getRepeatPasswordErrorText
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+
+
+class RegistrationViewModel(app: Application, private val performRegistrationUserCase: PerformRegistrationUserCase) : AndroidViewModel(app) {
+
+    private val _uiState = MutableStateFlow(RegistrationUiState())
+    val uiState: StateFlow<RegistrationUiState> = _uiState.asStateFlow()
+
+    private val _effect: MutableSharedFlow<RegistrationEffect> = MutableSharedFlow(extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
+    val effect: SharedFlow<RegistrationEffect> = _effect.asSharedFlow()
+
+    fun onRegisterClicked(name: String, lastName: String, email: String, password: String, repeatedPassword: String) {
+        validateName(name, false)
+        validateLastName(lastName, false)
+        validateEmail(email, false)
+        validatePass(password, false)
+        validateRepeatedPass(password, repeatedPassword, false)
+        if (!_uiState.value.hasAnyError()) {
+            viewModelScope.launch {
+                _uiState.value = RegistrationUiState(loading = true)
+                val result = performRegistrationUserCase(name, lastName, email, password)
+                result.onSuccess {
+                    _effect.tryEmit(RegistrationEffect.NavigateToEventList)
+                }.onFailure { error ->
+                    _uiState.value = _uiState.value.copy(loading = false)
+                    when {
+                        error is EsmorgaException && error.code == ErrorCodes.NO_CONNECTION -> _effect.tryEmit(RegistrationEffect.ShowNoNetworkSnackbar)
+                        error is EsmorgaException && error.code == 409 -> _uiState.value = RegistrationUiState(nameError = getEmailAlreadyInUseErrorText())
+                        else -> _effect.tryEmit(RegistrationEffect.ShowFullScreenError())
+                    }
+                }
+            }
+        }
+    }
+
+    fun validateName(name: String, acceptsEmpty: Boolean = true) {
```

Comment:
seems that all those fun are really similar, did you considered a way to avoid repeat all this code?

(AI)
```
fun validateField(
    fieldValue: String,
    fieldName: String,
    regex: String,
    errorTextProvider: () -> String,
    acceptsEmpty: Boolean
) {
    val isBlank = fieldValue.isBlank()
    val isValid = fieldValue.isEmpty() || fieldValue.matches(regex.toRegex())

    _uiState.value = when {
        !acceptsEmpty && isBlank -> _uiState.value.copy(
            errorTextProvider() to getEmptyFieldErrorText()
        )
        !isValid -> _uiState.value.copy(
            errorTextProvider() to getFieldErrorText(fieldName)
        )
        else -> _uiState.value.copy(
            errorTextProvider() to null
        )
    }
}
```